File: ctci/chapter_1_arrays_strings/js/1.6-stringCompression.js
/* O(n) solution */

/**
 * Compresses consecutive characters into a number and inserts it into a new
 * string along with character.
 * @param   {string}  str
 * @returns {string}
 */
const stringCompression = str => {
  /* String concatentation in JS is highly optimized. */
  result_str = ""
  let count = 1
  for (let i = 0; i < str.length; i++) {
    if (str[i] != str[i + 1]) {
      result_str += str[i] + count
      count = 1
    } else count++
  }
  return result_str
}

console.log(stringCompression("aaabbc")) // a3b2c1
console.log(stringCompression("abc")) // a1b1c1
console.log(stringCompression("lliiiinnnuux")) // l2i4n3u2x1

File: ctci/chapter_1_arrays_strings/1.6-string_compression.py
# O(n) solution


def string_compression(string: str) -> str:
    """
    Compresses consecutive characters into a number and inserts it into a new
    string along with character.

    String concatentation is slow because of immutability so using a list
    instead.
    """
    _list = []
    count = 1
    for i, e in enumerate(string):
        # if at last index or current and next chars don't match
        if i == len(string) - 1 or string[i] != string[i + 1]:
            _list += [e, str(count)]
            count = 1
        else:
            count += 1
    return ''.join(_list) if len(''.join(_list)) < len(string) else string


if __name__ == '__main__':
    print(string_compression("aaabbc"))  # a3b2c1
    print(string_compression("abc"))  # a1b1c1
    print(string_compression("lliiiinnnuux"))  # l2i4n3u2x1
